File: backend/app/schemas/v2/patient.py
"""PatientV2 / WeeklyPatternV2 schemas (Wave 0-C).

設計仕様書 v0.9 §4.1 に基づき、v2 で残す **16 項目** のみを保持する。
v1 で削除する 10 項目（年齢 / NG時間 / エリア / 指定タイプ / NGスタッフ /
同行希望スタッフ / 継続要望 / 必要スタッフ数 / 曜日優先度 / NG曜日）は
v2 schema には含めない。

追加項目（§3.3 / §3.4）:
- weekly_pattern.staff_count (既定 1, 2 で 2 名体制)
- special_weekly_pattern (JSONB)
- special_week_active (適用週リスト [{iso_year, iso_week}, ...])
"""

from __future__ import annotations

import re
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

# 7 weekday short codes used in JSONB (Mon..Sun, English short form).
# v2 でも v1 と同じ表現を維持する。
WeekdayV2 = Literal["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

# 時間タイプ (§4.1):
#   固定 / 時間帯 (preferred_start / preferred_end 必須)
#   午前 / 午後 / 終日 (時刻不要)
TimeTypeV2 = Literal["固定", "時間帯", "午前", "午後", "終日"]

# 訪問頻度 (§4.1 / §7.1 残置確定):
#   毎週 / 隔週 / 月次。具体的な使い方は別途検討（"TBD" のまま）。
VisitFrequencyV2 = Literal["every", "biweekly", "monthly"]

# 性別 (§4.1):
#   保存値は英語小文字。UI 表示は ja-JP に変換する。
SexV2 = Literal["male", "female", "unknown"]

# 性別制限 (§4.1, ハード制約):
#   "female_only" = 女性スタッフのみ可
#   "male_only" = 男性スタッフのみ可
#   None = 制限なし
SexRestrictionV2 = Literal["female_only", "male_only"]

# 患者状態 (§4.1):
#   active = 稼働中 (割当対象)
#   suspended = 一時休止
#   admitted = 入院中
#   pending = 開始前
#   cancelled = 解約済み
PatientStatusV2 = Literal["active", "suspended", "admitted", "pending", "cancelled"]

# 保険区分 (§4.1):
InsuranceV2 = Literal["medical", "care"]

_HHMM_RE = re.compile(r"^([01]\d|2[0-3]):[0-5]\d$")


def _validate_hhmm(value: str | None) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str) or not _HHMM_RE.match(value):
        raise ValueError(f"invalid HH:MM time string: {value!r}")
    return value


class WeeklyPatternEntryV2(BaseModel):
    """希望訪問パターンの 1 曜日エントリ (§3.3 / §4.1).

    `weekly_pattern` JSONB は曜日キー → このエントリ の形を取る場合と、
    リスト形式 (`[{weekday, ...}]`) の場合の両方が運用上ありうる。
    本クラスは後者の要素として用い、前者の場合は `WeeklyPatternV2.entries`
    に flatten したリストを格納する想定。
    """

    model_config = ConfigDict(extra="allow")

    weekday: WeekdayV2 | None = None
    time_type: TimeTypeV2 | None = None
    preferred_start: str | None = Field(
        default=None,
        description="HH:MM (24h). time_type=固定/時間帯 のときに使用",
    )
    preferred_end: str | None = Field(
        default=None,
        description="HH:MM (24h). time_type=時間帯 のときに使用",
    )
    service_minutes: int | None = Field(default=None, ge=0, le=300)
    staff_count: int = Field(
        default=1,
        ge=1,
        le=2,
        description="必要スタッフ数 (§3.3). 既定 1, 2 で 2 名体制",
    )

    @field_validator("preferred_start", "preferred_end")
    @classmethod
    def _check_hhmm(cls, v: str | None) -> str | None:
        return _validate_hhmm(v)


class WeeklyPatternV2(BaseModel):
    """患者の希望訪問パターン (§4.1 / §3.3).

    JSONB として `patients.weekly_pattern` に保存される。
    通常パターンと特別週パターン (`special_weekly_pattern`) で同じ shape を共有。
    """

    model_config = ConfigDict(extra="allow")

    frequency_per_week: int | None = Field(
        default=None,
        ge=1,
        le=7,
        description="週あたり訪問回数 (§4.1 核項目)",
    )
    visit_frequency: VisitFrequencyV2 | None = Field(
        default=None,
        description="訪問頻度 (毎週/隔週/月次, §7.1 残置確定, 具体仕様 TBD)",
    )
    visit_weeks: str | None = Field(
        default=None,
        description="訪問週 (例 '1,3', §7.1 残置確定, 具体仕様 TBD)",
    )
    preferred_weekdays: list[WeekdayV2] | None = None
    service_minutes: int | None = Field(default=None, ge=0, le=300)
    time_type: TimeTypeV2 | None = None
    preferred_start: str | None = None
    preferred_end: str | None = None
    entries: list[WeeklyPatternEntryV2] | None = Field(
        default=None,
        description=(
            "曜日ごとの細粒度設定 (§3.3 staff_count 等). "
            "リスト形式運用時に使用。曜日サマリ系は preferred_weekdays/time_type 等が真値"
        ),
    )

    @field_validator("preferred_start", "preferred_end")
    @classmethod
    def _check_hhmm(cls, v: str | None) -> str | None:
        return _validate_hhmm(v)


class SpecialWeekRefV2(BaseModel):
    """special_week_active の 1 要素 (§3.4 / §4.1).

    "適用週リスト [{iso_year, iso_week}, ...]" の各要素。
    """

    model_config = ConfigDict(extra="forbid")

    iso_year: int = Field(ge=2000, le=2100)
    iso_week: int = Field(ge=1, le=53)


class PatientV2Base(BaseModel):
    """患者マスタ v2 (§4.1 残す 16 項目).

    削除済み (v2 schema に含めない): age, ng_time_start, ng_time_end,
    required_staff_count, area, ng_staff_ids, preferred_staff_ids,
    specified_type, continuous_request, weekday_priority。
    """

    model_config = ConfigDict(extra="forbid")

    # 基本情報
    code: str = Field(min_length=1, max_length=64, description="患者コード (一意識別)")
    name: str = Field(min_length=1, max_length=120)
    kana: str | None = Field(default=None, max_length=120, description="表示・並び順")
    sex: SexV2 | None = Field(default=None, description="性別制限判定にも使用")
    status: PatientStatusV2 = Field(
        default="active",
        description="稼働/休止/入院/解約 等の割当除外フラグ",
    )

    # 連絡先
    address: str | None = Field(default=None, description="距離計算ベースの住所")
    lat: float | None = Field(default=None, ge=-90, le=90, description="距離計算用緯度")
    lng: float | None = Field(default=None, ge=-180, le=180, description="距離計算用経度")

    # 保険・拠点
    insurance: InsuranceV2 | None = Field(default=None, description="医療/介護")
    primary_office_id: UUID | None = Field(
        default=None,
        description=(
            "主担当拠点 (§4.3). 住所から自動紐付け + 手動上書き可。稲毛 / 都賀のみ運用 (MVP)"
        ),
    )

    # 訪問条件
    sex_restriction: SexRestrictionV2 | None = Field(
        default=None,
        description="ハード制約 (§4.1 / §5.4)",
    )
    # W18 Phase A: 複数スタッフ同行必須フラグ (§4.1 / §5.4).
    # 患者単位の独立フラグ。weekly_pattern.entries[].staff_count とは別軸。
    requires_multiple_staff: bool = Field(
        default=False,
        description=(
            "TRUE で常に 2 名体制が必要な患者. weekly_pattern.staff_count とは独立した "
            "患者単位フラグ (§4.1 / §5.4)."
        ),
    )

    # 希望訪問パターン (§4.1 核)
    weekly_pattern: WeeklyPatternV2 | None = Field(
        default=None,
        description="通常パターン (§3.4 / §4.1)",
    )
    special_weekly_pattern: WeeklyPatternV2 | None = Field(
        default=None,
        description="特別週用パターン (§3.4 Option A 確定)",
    )
    special_week_active: list[SpecialWeekRefV2] = Field(
        default_factory=list,
        description=(
            "適用週リスト [{iso_year, iso_week}, ...] (§3.4 / §4.1). "
            "該当週は special_weekly_pattern を優先、非該当週は weekly_pattern を使用"
        ),
    )

    # 備考
    note: str | None = Field(default=None, description="自由記述")


class PatientV2Create(PatientV2Base):
    """POST /api/v1/patients リクエスト (W1-BE1)."""


class PatientV2Update(BaseModel):
    """PATCH /api/v1/patients/{id} リクエスト (W1-BE1).

    全フィールド optional. base から `.partial()` 風に手書きで列挙する
    （v1 patient.py のスタイル踏襲。未指定値が default で誤って上書きされるのを防ぐため）。
    """

    model_config = ConfigDict(extra="forbid")

    code: str | None = Field(default=None, min_length=1, max_length=64)
    name: str | None = Field(default=None, min_length=1, max_length=120)
    kana: str | None = Field(default=None, max_length=120)
    sex: SexV2 | None = None
    status: PatientStatusV2 | None = None
    address: str | None = None
    lat: float | None = Field(default=None, ge=-90, le=90)
    lng: float | None = Field(default=None, ge=-180, le=180)
    insurance: InsuranceV2 | None = None
    primary_office_id: UUID | None = None
    sex_restriction: SexRestrictionV2 | None = None
    # W18 Phase A: 患者単位の複数スタッフ必須フラグ. None=未指定 (PATCH 時に touch しない)
    requires_multiple_staff: bool | None = None
    weekly_pattern: WeeklyPatternV2 | None = None
    special_weekly_pattern: WeeklyPatternV2 | None = None
    special_week_active: list[SpecialWeekRefV2] | None = None
    note: str | None = None


class PatientV2Read(PatientV2Base):
    """GET /api/v1/patients/{id} レスポンス (W1-BE1).

    Read 経路は JSONB を緩く受ける (v1 patient.py と同じ asymmetric 戦略)。
    weekly_pattern / special_weekly_pattern は **dict | None** で受け、
    フロント側で coerce する。
    """

    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: UUID
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None

    # asymmetric: 書込みは構造化 schema、読込は dict のまま
    weekly_pattern: dict | None = None  # type: ignore[assignment]
    special_weekly_pattern: dict | None = None  # type: ignore[assignment]
    special_week_active: list[dict] = []  # type: ignore[assignment]

File: frontend/lib/schemas/v2/patient.ts
/**
 * PatientV2 / WeeklyPatternV2 zod schemas (Wave 0-C).
 *
 * 設計仕様書 v0.9 §4.1 に基づき、v2 で残す **16 項目** のみを保持する。
 * v1 で削除する 10 項目 (年齢 / NG時間 / エリア / 指定タイプ / NGスタッフ /
 * 同行希望スタッフ / 継続要望 / 必要スタッフ数 / 曜日優先度 / NG曜日) は
 * v2 schema には含めない。
 *
 * 追加項目 (§3.3 / §3.4):
 *   - weekly_pattern.staff_count (既定 1, 2 で 2 名体制)
 *   - special_weekly_pattern (JSONB)
 *   - special_week_active (適用週リスト [{iso_year, iso_week}, ...])
 *
 * Backend `backend/app/schemas/v2/patient.py` の Pydantic schema と
 * **完全一致** させる。
 */
import { z } from 'zod';

// ─────────────────────────────────────────────────────────────────────────
// Constants & primitive enums
// ─────────────────────────────────────────────────────────────────────────

/** 7 weekday short codes (Mon..Sun, English short form). */
export const WEEKDAY_V2_VALUES = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'] as const;
export const weekdayV2Enum = z.enum(WEEKDAY_V2_VALUES);
export type WeekdayV2 = z.infer<typeof weekdayV2Enum>;

/**
 * 時間タイプ (§4.1):
 *   固定 / 時間帯 (preferred_start / preferred_end 必須)
 *   午前 / 午後 / 終日 (時刻不要)
 */
export const TIME_TYPE_V2_VALUES = ['固定', '時間帯', '午前', '午後', '終日'] as const;
export const timeTypeV2Enum = z.enum(TIME_TYPE_V2_VALUES);
export type TimeTypeV2 = z.infer<typeof timeTypeV2Enum>;

/**
 * 訪問頻度 (§4.1 / §7.1 残置確定): 毎週 / 隔週 / 月次。
 * 具体的な使い方は別途検討 (TBD)。
 */
export const VISIT_FREQUENCY_V2_VALUES = ['every', 'biweekly', 'monthly'] as const;
export const visitFrequencyV2Enum = z.enum(VISIT_FREQUENCY_V2_VALUES);
export type VisitFrequencyV2 = z.infer<typeof visitFrequencyV2Enum>;

/** 性別 (§4.1). 保存値は英語小文字。UI 表示は ja-JP に変換する。 */
export const SEX_V2_VALUES = ['male', 'female', 'unknown'] as const;
export const sexV2Enum = z.enum(SEX_V2_VALUES);
export type SexV2 = z.infer<typeof sexV2Enum>;

/**
 * 性別制限 (§4.1, ハード制約):
 *   "female_only" = 女性スタッフのみ可
 *   "male_only" = 男性スタッフのみ可
 *   null = 制限なし
 */
export const SEX_RESTRICTION_V2_VALUES = ['female_only', 'male_only'] as const;
export const sexRestrictionV2Enum = z.enum(SEX_RESTRICTION_V2_VALUES);
export type SexRestrictionV2 = z.infer<typeof sexRestrictionV2Enum>;

/**
 * 患者状態 (§4.1):
 *   active = 稼働中 (割当対象)
 *   suspended = 一時休止
 *   admitted = 入院中
 *   pending = 開始前
 *   cancelled = 解約済み
 */
export const PATIENT_STATUS_V2_VALUES = [
  'active',
  'suspended',
  'admitted',
  'pending',
  'cancelled',
] as const;
export const patientStatusV2Enum = z.enum(PATIENT_STATUS_V2_VALUES);
export type PatientStatusV2 = z.infer<typeof patientStatusV2Enum>;

/** 保険区分 (§4.1). */
export const INSURANCE_V2_VALUES = ['medical', 'care'] as const;
export const insuranceV2Enum = z.enum(INSURANCE_V2_VALUES);
export type InsuranceV2 = z.infer<typeof insuranceV2Enum>;

// ─────────────────────────────────────────────────────────────────────────
// HH:MM validator
// ─────────────────────────────────────────────────────────────────────────
const hhmmRegex = /^([01]\d|2[0-3]):[0-5]\d$/;
const hhmmSchema = z.string().regex(hhmmRegex, '時刻は HH:MM 形式で入力してください');

// ─────────────────────────────────────────────────────────────────────────
// WeeklyPattern (§3.3 / §4.1)
// ─────────────────────────────────────────────────────────────────────────

/** 希望訪問パターンの 1 曜日エントリ (§3.3 / §4.1). */
export const weeklyPatternEntryV2Schema = z
  .object({
    weekday: weekdayV2Enum.nullable().optional(),
    time_type: timeTypeV2Enum.nullable().optional(),
    preferred_start: hhmmSchema.nullable().optional(),
    preferred_end: hhmmSchema.nullable().optional(),
    service_minutes: z.number().int().min(0).max(300).nullable().optional(),
    /** 必要スタッフ数 (§3.3). 既定 1, 2 で 2 名体制 */
    staff_count: z.number().int().min(1).max(2).default(1),
  })
  .passthrough();
export type WeeklyPatternEntryV2 = z.infer<typeof weeklyPatternEntryV2Schema>;

/**
 * 患者の希望訪問パターン (§4.1 / §3.3).
 *
 * JSONB として `patients.weekly_pattern` に保存される。
 * 通常パターンと特別週パターン (`special_weekly_pattern`) で同じ shape を共有。
 */
export const weeklyPatternV2Schema = z
  .object({
    /** 週あたり訪問回数 (§4.1 核項目) */
    frequency_per_week: z.number().int().min(1).max(7).nullable().optional(),
    /** 訪問頻度 (§7.1 残置確定, 具体仕様 TBD) */
    visit_frequency: visitFrequencyV2Enum.nullable().optional(),
    /** 訪問週 (例 "1,3", §7.1 残置確定, 具体仕様 TBD) */
    visit_weeks: z.string().nullable().optional(),
    preferred_weekdays: z.array(weekdayV2Enum).nullable().optional(),
    service_minutes: z.number().int().min(0).max(300).nullable().optional(),
    time_type: timeTypeV2Enum.nullable().optional(),
    preferred_start: hhmmSchema.nullable().optional(),
    preferred_end: hhmmSchema.nullable().optional(),
    /** 曜日ごとの細粒度設定 (§3.3 staff_count 等). リスト形式運用時に使用 */
    entries: z.array(weeklyPatternEntryV2Schema).nullable().optional(),
  })
  .passthrough();
export type WeeklyPatternV2 = z.infer<typeof weeklyPatternV2Schema>;

/** special_week_active の 1 要素 (§3.4 / §4.1). */
export const specialWeekRefV2Schema = z.object({
  iso_year: z.number().int().min(2000).max(2100),
  iso_week: z.number().int().min(1).max(53),
});
export type SpecialWeekRefV2 = z.infer<typeof specialWeekRefV2Schema>;

// ─────────────────────────────────────────────────────────────────────────
// PatientV2 base / Create / Update / Read
// ─────────────────────────────────────────────────────────────────────────

/**
 * 患者マスタ v2 (§4.1 残す 16 項目).
 *
 * 削除済み (v2 schema に含めない): age, ng_time_start, ng_time_end,
 * required_staff_count, area, ng_staff_ids, preferred_staff_ids,
 * specified_type, continuous_request, weekday_priority。
 */
export const patientV2BaseSchema = z.object({
  // 基本情報
  code: z.string().min(1, 'コードは必須です').max(64),
  name: z.string().min(1, '氏名は必須です').max(120),
  kana: z.string().max(120).nullable().optional(),
  sex: sexV2Enum.nullable().optional(),
  status: patientStatusV2Enum.default('active'),

  // 連絡先
  address: z.string().nullable().optional(),
  lat: z.number().min(-90).max(90).nullable().optional(),
  lng: z.number().min(-180).max(180).nullable().optional(),

  // 保険・拠点
  insurance: insuranceV2Enum.nullable().optional(),
  primary_office_id: z.string().uuid().nullable().optional(),

  // 訪問条件 (ハード制約)
  sex_restriction: sexRestrictionV2Enum.nullable().optional(),

  // 希望訪問パターン (§4.1 核)
  weekly_pattern: weeklyPatternV2Schema.nullable().optional(),
  special_weekly_pattern: weeklyPatternV2Schema.nullable().optional(),
  special_week_active: z.array(specialWeekRefV2Schema).default([]),

  // 備考
  note: z.string().nullable().optional(),
});
export type PatientV2Base = z.infer<typeof patientV2BaseSchema>;

/** POST /api/v1/patients リクエスト (W1-BE1). */
export const patientV2CreateSchema = patientV2BaseSchema;
export type PatientV2Create = z.infer<typeof patientV2CreateSchema>;

/**
 * PATCH /api/v1/patients/{id} リクエスト (W1-BE1).
 *
 * 全フィールド optional. 手書き列挙 (`.partial()` 不使用)。
 * デフォルト値が PATCH ペイロードに混入してサーバ値を黙って上書きするのを防ぐため。
 */
export const patientV2UpdateSchema = z.object({
  code: z.string().min(1).max(64).optional(),
  name: z.string().min(1).max(120).optional(),
  kana: z.string().max(120).nullable().optional(),
  sex: sexV2Enum.nullable().optional(),
  status: patientStatusV2Enum.optional(),
  address: z.string().nullable().optional(),
  lat: z.number().min(-90).max(90).nullable().optional(),
  lng: z.number().min(-180).max(180).nullable().optional(),
  insurance: insuranceV2Enum.nullable().optional(),
  primary_office_id: z.string().uuid().nullable().optional(),
  sex_restriction: sexRestrictionV2Enum.nullable().optional(),
  weekly_pattern: weeklyPatternV2Schema.nullable().optional(),
  special_weekly_pattern: weeklyPatternV2Schema.nullable().optional(),
  special_week_active: z.array(specialWeekRefV2Schema).optional(),
  note: z.string().nullable().optional(),
});
export type PatientV2Update = z.infer<typeof patientV2UpdateSchema>;

/**
 * GET /api/v1/patients/{id} レスポンス (W1-BE1).
 *
 * Read 経路は JSONB を緩く受ける (asymmetric 戦略: 書込みは構造化、読込は dict)。
 * weekly_pattern / special_weekly_pattern は dict のまま受け、
 * `coerceWeeklyPattern` 等で UI 用に変換する。
 */
export const patientV2ReadSchema = patientV2BaseSchema.extend({
  id: z.string().uuid(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional(),
  // asymmetric: 読込は dict のまま受ける
  weekly_pattern: z.record(z.unknown()).nullable().optional(),
  special_weekly_pattern: z.record(z.unknown()).nullable().optional(),
  special_week_active: z.array(z.record(z.unknown())).default([]),
});
export type PatientV2Read = z.infer<typeof patientV2ReadSchema>;
